File: CBTReframe/Services/AnthropicService.swift
import Foundation

struct AnthropicService: AIServiceProtocol {
    let provider = AIProvider.anthropic

    func reframe(
        thought: String,
        mood: String,
        model: AIModel,
        mode: ReframeMode,
        style: ResponseStyle,
        template: PromptTemplate,
        strategy: ResponseStrategy
    ) async throws -> AnalysisResult {
        guard let apiKey = KeychainManager.shared.load(key: provider.rawValue),
              !apiKey.isEmpty else {
            throw AIServiceError.noAPIKey
        }

        let systemPrompt = PromptBuilder.buildSystemPrompt(mode: mode, style: style, template: template, strategy: strategy)
        let userPrompt = PromptBuilder.buildUserPrompt(thought: thought, mood: mood)

        let body: [String: Any] = [
            "model": model.id,
            "max_tokens": strategy == .crisis ? 512 : 1024,
            "system": systemPrompt,
            "messages": [
                ["role": "user", "content": userPrompt],
            ],
        ]

        var request = URLRequest(url: URL(string: provider.baseURL)!)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.timeoutInterval = 60

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIServiceError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200: break
        case 401: throw AIServiceError.invalidKey
        case 429: throw AIServiceError.rateLimited
        default: throw AIServiceError.invalidResponse
        }

        return try parseAnthropicResponse(data, strategy: strategy)
    }

    func analyzeThoughtPatterns(
        thoughts: [ThoughtEntry],
        model: AIModel
    ) async throws -> ThoughtPatternReport {
        guard let apiKey = KeychainManager.shared.load(key: provider.rawValue),
              !apiKey.isEmpty else {
            throw AIServiceError.noAPIKey
        }

        let body: [String: Any] = [
            "model": model.id,
            "max_tokens": 1400,
            "system": PromptBuilder.thoughtPatternSystemPrompt,
            "messages": [
                ["role": "user", "content": PromptBuilder.buildThoughtPatternUserPrompt(thoughts: thoughts)],
            ],
        ]

        var request = URLRequest(url: URL(string: provider.baseURL)!)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.timeoutInterval = 60

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIServiceError.invalidResponse
        }

        switch httpResponse.statusCode {
        case 200: break
        case 401: throw AIServiceError.invalidKey
        case 429: throw AIServiceError.rateLimited
        default: throw AIServiceError.invalidResponse
        }

        return try parseAnthropicThoughtPatternResponse(data)
    }

    private func parseAnthropicResponse(_ data: Data, strategy: ResponseStrategy) throws -> AnalysisResult {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let content = json["content"] as? [[String: Any]],
              let textBlock = content.first(where: { $0["type"] as? String == "text" }),
              let text = textBlock["text"] as? String else {
            throw AIServiceError.invalidResponse
        }

        return try parseReframeOutput(text, strategy: strategy)
    }

    private func parseAnthropicThoughtPatternResponse(_ data: Data) throws -> ThoughtPatternReport {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let content = json["content"] as? [[String: Any]],
              let textBlock = content.first(where: { $0["type"] as? String == "text" }),
              let text = textBlock["text"] as? String else {
            throw AIServiceError.invalidResponse
        }

        return try parseThoughtPatternContent(text)
    }
}

File: shared/src/commonMain/kotlin/com/henryliu/cbtreframe/shared/AnthropicService.kt
package com.henryliu.cbtreframe.shared

import io.ktor.client.HttpClient
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerialName
import kotlinx.serialization.json.Json

@Serializable
data class AnthropicBody(
    val model: String,
    @SerialName("max_tokens")
    val maxTokens: Int,
    val system: String,
    val messages: List<AnthropicMessage>,
)

@Serializable
data class AnthropicMessage(
    val role: String,
    val content: String,
)

@Serializable
data class AnthropicContentBlock(
    val type: String,
    val text: String? = null,
)

@Serializable
data class AnthropicResponse(
    val content: List<AnthropicContentBlock> = emptyList(),
)

class AnthropicService(
    private val httpClient: HttpClient,
    private val apiKeyProvider: suspend () -> String?,
) : AIServiceProtocol {
    override val provider: AIProvider = AIProvider.ANTHROPIC
    private val baseUrl = "https://api.anthropic.com/v1/messages"

    override suspend fun reframe(
        thought: String,
        mood: String,
        hasAkathisia: Boolean,
        model: AIModel,
        depth: ThinkingTemplate.AnalysisDepth,
        style: ThinkingTemplate.AppResponseStyle,
        template: ThinkingTemplate,
        strategy: ResponseStrategy,
    ): AnalysisResult {
        val apiKey = apiKeyProvider() ?: throw AIServiceError.NoAPIKey()

        val systemPrompt = PromptBuilder.buildSystemPrompt(
            template = template,
            strategy = strategy,
            depth = depth,
            style = style,
            mood = mood,
            hasAkathisia = hasAkathisia,
        )
        val userPrompt = PromptBuilder.buildUserPrompt(thought, mood, hasAkathisia)

        val body = AnthropicBody(
            model = model.modelName,
            maxTokens = if (strategy == ResponseStrategy.crisis) 512 else 1024,
            system = systemPrompt,
            messages = listOf(
                AnthropicMessage(role = "user", content = userPrompt),
            ),
        )

        val response = httpClient.post(baseUrl) {
            header("x-api-key", apiKey)
            header("anthropic-version", "2023-06-01")
            contentType(ContentType.Application.Json)
            setBody(anthropicJson.encodeToString(AnthropicBody.serializer(), body))
        }

        when (response.status.value) {
            200 -> { /* ok */ }
            401 -> throw AIServiceError.InvalidKey()
            429 -> throw AIServiceError.RateLimited()
            else -> throw AIServiceError.InvalidResponse()
        }

        val data = response.bodyAsText()
        return parseAnthropicResponse(data, strategy)
    }

    override suspend fun analyzeThoughtPatterns(
        thoughts: List<ThoughtEntry>,
        model: AIModel,
    ): ThoughtPatternReport {
        val apiKey = apiKeyProvider() ?: throw AIServiceError.NoAPIKey()

        val body = AnthropicBody(
            model = model.modelName,
            maxTokens = 1400,
            system = PromptBuilder.thoughtPatternSystemPrompt,
            messages = listOf(
                AnthropicMessage(
                    role = "user",
                    content = PromptBuilder.buildThoughtPatternUserPrompt(thoughts),
                ),
            ),
        )

        val response = httpClient.post(baseUrl) {
            header("x-api-key", apiKey)
            header("anthropic-version", "2023-06-01")
            contentType(ContentType.Application.Json)
            setBody(anthropicJson.encodeToString(AnthropicBody.serializer(), body))
        }

        when (response.status.value) {
            200 -> { /* ok */ }
            401 -> throw AIServiceError.InvalidKey()
            429 -> throw AIServiceError.RateLimited()
            else -> throw AIServiceError.InvalidResponse()
        }

        val data = response.bodyAsText()
        return parseAnthropicThoughtPatternResponse(data)
    }

    private fun parseAnthropicResponse(data: String, strategy: ResponseStrategy): AnalysisResult {
        val resp = parseAnthropicResponseJson(data) ?: throw AIServiceError.InvalidResponse()
        val textBlock = resp.content.firstOrNull { it.type == "text" }
            ?: throw AIServiceError.InvalidResponse()
        val text = textBlock.text ?: throw AIServiceError.InvalidResponse()
        return parseReframeOutput(text, strategy)
    }

    private fun parseAnthropicThoughtPatternResponse(data: String): ThoughtPatternReport {
        val resp = parseAnthropicResponseJson(data) ?: throw AIServiceError.InvalidResponse()
        val textBlock = resp.content.firstOrNull { it.type == "text" }
            ?: throw AIServiceError.InvalidResponse()
        val text = textBlock.text ?: throw AIServiceError.InvalidResponse()
        return parseThoughtPatternContent(text)
    }

    private fun parseAnthropicResponseJson(data: String): AnthropicResponse? {
        return try {
            anthropicJson.decodeFromString(AnthropicResponse.serializer(), data)
        } catch (_: Exception) {
            null
        }
    }
}

private val anthropicJson = Json { ignoreUnknownKeys = true; isLenient = true }
